File: tests/machine/x/clj/list_assign.clj
(ns main)

(defn _indexList [xs i]
  (let [idx (if (neg? i) (+ i (count xs)) i)]
    (if (or (< idx 0) (>= idx (count xs)))
      (throw (ex-info "index out of range" {}))
      (nth xs idx))))

(declare nums)

(defn -main []
  (def nums [1 2]) ;; list of int
  (def nums (assoc nums 1 3)) ;; int
  (println (_indexList nums 1))
)

(-main)

File: tests/machine/x/cs/list_assign.cs
using System;
using System.Collections.Generic;

class Program {
    static void Main() {
        List<int> nums = new List<int> { 1, 2 };
        nums[1] = 3;
        Console.WriteLine(_indexList(nums, 1));
    }
    static dynamic _indexList(dynamic l, long i) {
        var list = l as System.Collections.IList;
        if (list == null) throw new Exception("index() expects list");
        if (i < 0) i += list.Count;
        if (i < 0 || i >= list.Count) throw new Exception("index out of range");
        return list[(int)i];
    }
    
}
